source/agora/consensus/protocol/Nominator.d

```diff
@@ -817,45 +817,58 @@ extern(D):
     private bool collectBlockSignature (in ValidatorBlockSig block_sig,
         in Hash block_hash) @safe nothrow
     {
-        const PublicKey K = block_sig.public_key;
         auto sigs = block_sig.height in this.slot_sigs;
-        if (block_sig.height in this.slot_sigs && K in this.slot_sigs[block_sig.height])
+        if (block_sig.height in this.slot_sigs && block_sig.utxo in this.slot_sigs[block_sig.height])
         {
             log.trace("Signature already collected for this node at height {}", block_sig.height);
             return false;
         }
-        if (!K.isValid())
+        ValidatorInfo validator;
+        // FIXME: Linear search should be replaced by amortized O(1)
```

Comment:
BTW, Is it possible to get O(1)? The most efficient overhead is O(n), Isn't it?

Reply:
An AA is amortized `O(1)`, and since the array is sorted, we can do a binary search to get `O(log(n))`.
I'll fix the comment.

Reply:
Went with `assumeSorted` and `find`.